doc/design/proposal-noderc.md

```diff
@@ -0,0 +1,164 @@
+
+```json
+{
+  "schema": 0,
+  "import": [ { "specifier": "./monitor.js", "mainThreadOnly": false } ]
+}
+```
+
+It also allows nicer-looking sub-configurations without repeating the prefix required by CLI flags, for example:
+
+```json
+{
+  "schema": 0,
+  "test": {
+    "reporter": "tap",
+    "name-pattern": [ "test [1-3]", "/test [4-5]/i" ]
+  }
+}
+```
+
+### Overriding the noderc file being applied
+
+We should provide a way for users to override the noderc file that's discoverd through `package.json` lookups, or to disable it. This can be done through either environment variables, or CLI flags, or both.
+
+The environment variable/CLI flag can specify a registered rc file in `package.json`, where `"noderc"` contains key-value pairs instead of just a string:
+
+```json
+{
+  "noderc": {
+    "default": "./.noderc.json",
+    "test": "./.noderc.test.json",
+    "test:coverage": "./.noderc.test.coverage.json"
+  }
+}
+```
+
+Suppose the environment variable is called `NODE_RC`, then `NODE_RC=test` results in the resolution of `./.noderc.test.json`. If `NODE_RC` is unspecified, and the `"noderc"` field in `package.json` contains key-value pairs, the rc file pointed by `"default"` will be selected by default.
+
+Open question: what should be the key of the default noderc? `default` might still class with potential `script` fields, which may or may not be a problem in the script -> noderc mapping we'll discuss below.
+
+### Accompanying JS APIs
+
+This feature should have some accompanying JS APIs to:
+
+1. Parse a given rc file
+2. Serialize a given configuration
+3. Querying the current configuration applied to the process (with an option to include additional configurations added by CLI flags/environment variables/runtime APIs), and where they come from
+4. Querying the current schema schema, and the features supported
+
+IDEs and other tooling are expected to use a matching Node.js version (likely specified by the "engine" field in the applicable `package.json`) to modify the rc files.
+
+This will be left to later iterations.
+
+### Reusing/extending configurations
+
+A `noderc` file can extend other `noderc` files. For example, for the main `noderc` pointed by `package.json`, it has an import that enables TypeScript loading:
+
+```json
+{
+  "schema": 0,
+  "import": [ "amaro/register" ]
+}
+```
+
+In addition, there can be another `./.noderc.prod.json` extending it for starting the server in production with monitor on:
+
+```json
+{
+  "schema": 0,
+  "extends": [ "./.noderc.json" ],
+  "import": [ "./monitor.ts" ]
+}
+```
+
+Open question: how should we handle override v.s. concatenation? Should we invent a special syntax? For example, to concatenate, use `"+import"`, otherwise, use `"import"`? Or is that too cryptic and we should just do "override if it's not an array, concatenate if it's an array?"
```

Comment:
If "extend" field is an array, meaning it can extend multiple files, I think its important to allow more granular configuration to chose which extend and which override.
This could be achieved by 

```json
{
"extend" : [{
 "alias": "foo",
 "path": "./.noderc.foo.json"
}, "./.noderc.bar.json"],
"import": [{ "alias": "foo" }, "amaro/register"]
}
```

This could allow to esplicitly estend `foo` and override others. (Bikeshed on naming etc...)

Reply:
What is the `alias` field specifying here? Is it some kind of variable-like thing?

We could also just provide `"override"` either as a different array, or do a filtering of configs in `extend`

```
"extend": [{ "filter": ["import"], "path": "./.noderc.foo.json" }]
```